sw/host/opentitanlib/src/crypto/rsa.rs

```diff
@@ -34,33 +34,41 @@ fixed_size_bigint!(N0Inv, at_most OTBN_BITS);
 #[derive(Debug, Error)]
 pub enum Error {
     #[error("Invalid public key")]
-    InvalidPublicKey,
+    InvalidPublicKey(#[source] Option<anyhow::Error>),
     #[error("Invalid DER file: {der}")]
     InvalidDerFile {
         der: PathBuf,
         #[source]
         source: anyhow::Error,
     },
-    #[error("Read failed: {0}")]
-    ReadFailed(PathBuf),
-    #[error("Write failed: {0}")]
-    WriteFailed(PathBuf),
+    #[error("Read failed: {file}")]
+    ReadFailed {
+        file: PathBuf,
+        #[source]
+        source: anyhow::Error,
+    },
+    #[error("Write failed: {file}")]
+    WriteFailed {
+        file: PathBuf,
+        #[source]
+        source: anyhow::Error,
+    },
     #[error("Generate failed")]
-    GenerateFailed,
+    GenerateFailed(#[source] anyhow::Error),
     #[error("Invalid signature")]
-    InvalidSignature,
+    InvalidSignature(#[source] anyhow::Error),
     #[error("Sign failed")]
-    SignFailed,
+    SignFailed(#[source] anyhow::Error),
     #[error("Verification failed")]
-    VerifyFailed,
+    VerifyFailed(#[source] anyhow::Error),
```

Comment:
I'm not sure, there are some `rsa::pkcs8::Error`s and `std::io::Error`s in there. Maybe some of these can be changed to their proper type, but I figured this was simpler. Are there any downsides to using `anyhow!()` apart from the extra macro invocation?

Reply:
Are the additional enum variants providing any value?  The `anyhow::Result` type that all of our functions use already converts all error types into `anyhow::Error`, so wrapping them inside of these enum variants my not be all that helpful (and we can avoid explicit thunks through `anyhow!()` by turning on feature `std` in the `rsa` crate).

If we want to supply additional context about errors, then `use anyhow::Context` and supply the information at the error-creating call site:

```
self.key
    .write_public_key_der_file(&der_file)
    .context(format!("Writing DER file: {:?}", der_file))?
```

This will emit an error message like:
```
Error: Writing DER file: foobar.der

Caused by:
    <the display form of the actual error>
```

Reply:
BTW, the update of the RSA crate for `hsmtool` turns on `std` by default in that crate.  The thunks through `anyhow!` should no longer be needed.